// ngx-editor/esm5/lib/ngx-editor.component.js
import { __decorate } from "tslib";
import { Component, ViewChild, forwardRef, ViewEncapsulation } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { EditorState } from 'prosemirror-state';
import { EditorView } from 'prosemirror-view';
import { NgxEditorService, NgxEditorServiceConfig } from './ngx-editor.service';
import schema from './schema';
var NgxEditorComponent = /** @class */ (function () {
    function NgxEditorComponent(ngxEditorService) {
        this.config = ngxEditorService.config;
    }
    NgxEditorComponent_1 = NgxEditorComponent;
    NgxEditorComponent.prototype.writeValue = function (value) {
        if (!value) {
            return;
        }
        this.updateContent(value);
    };
    NgxEditorComponent.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    NgxEditorComponent.prototype.registerOnTouched = function () { };
    NgxEditorComponent.prototype.parseDoc = function (contentJson) {
        return schema.nodeFromJSON(contentJson);
    };
    NgxEditorComponent.prototype.updateContent = function (value) {
        try {
            var doc = this.parseDoc(value);
            var state = this.view.state;
            var tr = state.tr;
            tr.replaceWith(0, state.doc.content.size, doc);
            this.view.dispatch(tr);
        }
        catch (err) {
            console.error('Unable to update document.', err);
        }
    };
    NgxEditorComponent.prototype.handleTransactions = function (tr) {
        var state = this.view.state.applyTransaction(tr).state;
        this.view.updateState(state);
        if (tr.docChanged) {
            var json = state.doc.toJSON();
            this.onChange(json);
        }
    };
    NgxEditorComponent.prototype.createEditor = function () {
        this.view = new EditorView(this.ngxEditor.nativeElement, {
            state: EditorState.create({
                schema: schema,
                plugins: this.config.plugins,
            }),
            dispatchTransaction: this.handleTransactions.bind(this),
            attributes: {
                class: 'NgxEditor-Content'
            },
        });
    };
    NgxEditorComponent.prototype.ngOnInit = function () {
        this.createEditor();
    };
    NgxEditorComponent.prototype.ngOnDestroy = function () {
        this.view.destroy();
    };
    var NgxEditorComponent_1;
    NgxEditorComponent.ctorParameters = function () { return [
        { type: NgxEditorService }
    ]; };
    __decorate([
        ViewChild('ngxEditor', { static: true })
    ], NgxEditorComponent.prototype, "ngxEditor", void 0);
    NgxEditorComponent = NgxEditorComponent_1 = __decorate([
        Component({
            selector: 'ngx-editor',
            template: "<div class=\"NgxEditor-Wrapper\" #ngxEditor></div>\n",
            providers: [{
                    provide: NG_VALUE_ACCESSOR,
                    useExisting: forwardRef(function () { return NgxEditorComponent_1; }),
                    multi: true
                }],
            encapsulation: ViewEncapsulation.None,
            styles: [".ProseMirror{outline:0}.NgxEditor-Wrapper{background:padding-box #fff;color:#000;border-radius:4px;border:2px solid rgba(0,0,0,.2)}.NgxEditor-MenuBar{display:flex;padding:.2rem;border-bottom:1px solid #ddd}.NgxEditor-MenuBar .NgxEditor-MenuItem{display:flex;align-items:center;justify-content:center}.NgxEditor-MenuBar .NgxEditor-MenuItem:hover{cursor:pointer;background-color:#f1f1f1}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Icon{height:30px;width:30px;transition:.3s ease-in-out;border-radius:2px;margin-right:2px}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Icon.NgxEditor-MenuItem__Active{background-color:#e8f0fe}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Icon.NgxEditor-MenuItem__Active svg{fill:#1a73e8}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper{position:relative;max-width:100px;width:100%;justify-content:start}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper .NgxEditor-MenuItem__Dropdown{display:flex;align-items:center;width:100%}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper .NgxEditor-MenuItem__Dropdown .NgxEditor-MenuItem__Dropdown-Text{margin-left:5px}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper .NgxEditor-MenuItem__Dropdown .NgxEditor-MenuItem__Dropdown-Icon{margin-left:auto;position:relative;top:2px}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper .NgxEditor-MenuItem__Dropdown-Menu{position:absolute;left:0;top:30px;box-shadow:rgba(60,64,67,.15) 0 2px 6px 2px;border-radius:4px;background-color:#fff;display:none}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper .NgxEditor-MenuItem__Dropdown-Menu .NgxEditor-MenuItem__Dropdown-Item{cursor:pointer;padding:.5rem 1rem;white-space:nowrap}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper .NgxEditor-MenuItem__Dropdown-Menu .NgxEditor-MenuItem__Dropdown-Item.NgxEditor-MenuItem__active,.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper .NgxEditor-MenuItem__Dropdown-Menu .NgxEditor-MenuItem__Dropdown-Item:hover{background-color:#f1f1f1}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper.NgxEditor-MenuItem__Dropdown-Wrapper-Open,.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper.NgxEditor-MenuItem__Dropdown-Wrapper-Selected{background-color:#e8f0fe;color:#1a73e8}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper.NgxEditor-MenuItem__Dropdown-Wrapper-Open .NgxEditor-MenuItem__Dropdown-Icon svg{fill:#1a73e8}.NgxEditor-MenuBar .NgxEditor-MenuItem.NgxEditor-MenuItem__Dropdown-Wrapper.NgxEditor-MenuItem__Dropdown-Wrapper-Open .NgxEditor-MenuItem__Dropdown-Menu{display:block;color:initial}.NgxEditor-MenuBar .NgxEditor-MenuItem__Seperator{border-left:1px solid #ccc;margin:0 5px}.NgxEditor-Content{padding:.5rem;white-space:pre-wrap}.NgxEditor-Content p{margin:0 0 .7rem}.NgxEditor-Placeholder{color:#6c757d;opacity:1}"]
        })
    ], NgxEditorComponent);
    return NgxEditorComponent;
}());
export { NgxEditorComponent };
