// intl-input-phone/esm2015/lib/model/Enums.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const ContentOptionsEnum = {
    Flag: 0,
    CountryName: 1,
    CountryPhoneCode: 2,
};
export { ContentOptionsEnum };
ContentOptionsEnum[ContentOptionsEnum.Flag] = 'Flag';
ContentOptionsEnum[ContentOptionsEnum.CountryName] = 'CountryName';
ContentOptionsEnum[ContentOptionsEnum.CountryPhoneCode] = 'CountryPhoneCode';
/** @enum {number} */
const OutputOptionsEnum = {
    NumberWithCountryCode: 0,
    Number: 1,
};
export { OutputOptionsEnum };
OutputOptionsEnum[OutputOptionsEnum.NumberWithCountryCode] = 'NumberWithCountryCode';
OutputOptionsEnum[OutputOptionsEnum.Number] = 'Number';
/** @enum {number} */
const SortOrderEnum = {
    CountryName: 0,
    CountryISOCode: 1,
    CountryPhoneCode: 2,
};
export { SortOrderEnum };
SortOrderEnum[SortOrderEnum.CountryName] = 'CountryName';
SortOrderEnum[SortOrderEnum.CountryISOCode] = 'CountryISOCode';
SortOrderEnum[SortOrderEnum.CountryPhoneCode] = 'CountryPhoneCode';
/** @enum {number} */
const TooltipOptionsEnum = {
    CountryName: 1,
    CountryPhoneCode: 2,
    CountryISOCode: 3,
};
export { TooltipOptionsEnum };
TooltipOptionsEnum[TooltipOptionsEnum.CountryName] = 'CountryName';
TooltipOptionsEnum[TooltipOptionsEnum.CountryPhoneCode] = 'CountryPhoneCode';
TooltipOptionsEnum[TooltipOptionsEnum.CountryISOCode] = 'CountryISOCode';
