// intl-input-phone/esm5/lib/model/Configuration.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentOptionsEnum, SortOrderEnum, TooltipOptionsEnum, OutputOptionsEnum } from "./Enums";
import { Injectable } from '@angular/core';
/**
 * Class for configure the dropdown.
 */
var ConfigurationOptions = /** @class */ (function () {
    function ConfigurationOptions() {
        /**
         * Prperty to set the selector for the dropdown.
         * Default : 'IntlPhoneNumber'
         */
        this.SelectorClass = "IntlPhoneNumber";
        /**
         * Property to set which content show in option
         * Default : Flag, CountryPhoneCode
         */
        this.OptionTextTypes = [
            ContentOptionsEnum.Flag,
            ContentOptionsEnum.CountryPhoneCode
        ];
        /**
         * Property to set the tooltip text
         * Default : Country Name
         */
        this.ToolTipText = TooltipOptionsEnum.CountryName;
        /**
         * Property to set whether tooltip is show or not.
         * Default : true
         */
        this.IsShowToolTip = true;
        /**
         * Property to set whether user want search option or not.
         * Default : true
         */
        this.IsShowSearchOption = true;
        /**
         * Property to set the sort order of country list to be rendered.
         * Default : CountryISOCode
         */
        this.SortBy = SortOrderEnum.CountryISOCode;
        /**
         * Property to set whether show all other country in dropdown as well or not, when specify custom country list in module.
         * Default : true
         */
        this.IsShowAllOtherCountry = true;
        /**
         * Property to set in which format user want the output when fill the input correctly.
         * Default : NumberWithCountryCode
         */
        this.OutputFormat = OutputOptionsEnum.NumberWithCountryCode;
    }
    ConfigurationOptions.decorators = [
        { type: Injectable }
    ];
    return ConfigurationOptions;
}());
export { ConfigurationOptions };
if (false) {
    /**
     * Prperty to set the selector for the dropdown.
     * Default : 'IntlPhoneNumber'
     * @type {?}
     */
    ConfigurationOptions.prototype.SelectorClass;
    /**
     * Property to set which content show in option
     * Default : Flag, CountryPhoneCode
     * @type {?}
     */
    ConfigurationOptions.prototype.OptionTextTypes;
    /**
     * Property to set the tooltip text
     * Default : Country Name
     * @type {?}
     */
    ConfigurationOptions.prototype.ToolTipText;
    /**
     * Property to set whether tooltip is show or not.
     * Default : true
     * @type {?}
     */
    ConfigurationOptions.prototype.IsShowToolTip;
    /**
     * Property to set whether user want search option or not.
     * Default : true
     * @type {?}
     */
    ConfigurationOptions.prototype.IsShowSearchOption;
    /**
     * Property to set the sort order of country list to be rendered.
     * Default : CountryISOCode
     * @type {?}
     */
    ConfigurationOptions.prototype.SortBy;
    /**
     * Property to set whether show all other country in dropdown as well or not, when specify custom country list in module.
     * Default : true
     * @type {?}
     */
    ConfigurationOptions.prototype.IsShowAllOtherCountry;
    /**
     * Property to set in which format user want the output when fill the input correctly.
     * Default : NumberWithCountryCode
     * @type {?}
     */
    ConfigurationOptions.prototype.OutputFormat;
}
